optimism/src/keccak/interpreter.rs

```diff
@@ -124,219 +124,302 @@ pub trait KeccakInterpreter<F: One + Debug + Zero> {
+        // - 30 constraints of degree 2
+        // - 5 constraints of degree 3
+        let state_e = self.constrain_theta();
+
+        // STEP pirho: 5 * 5 * (2 + 4 * 1) = 150 constraints of degree 2
+        let state_b = self.constrain_pirho(state_e);
+
+        // STEP chi: 4 * 5 * 5 * 2 = 200 constraints of degree 2
+        let state_f = self.constrain_chi(state_b);
+
+        // STEP iota: 4 constraints of degree 2
+        self.constrain_iota(state_f);
+    }
+
+    /// Constrains 35 checks of the theta algorithm in round steps
+    ///  - 30 constraints of degree 2
+    ///  - 5 constraints of degree 3
+    // TODO: when circuits are split into Round and Sponge, these constraints will have 1 less degree
+    fn constrain_theta(&mut self) -> Vec<Vec<Vec<Self::Variable>>> {
+        // Define vectors storing expressions which are not in the witness layout for efficiency
+        let mut state_c = vec![vec![Self::zero(); QUARTERS]; DIM];
+        let mut state_d = vec![vec![Self::zero(); QUARTERS]; DIM];
+        let mut state_e = vec![vec![vec![Self::zero(); QUARTERS]; DIM]; DIM];
+
+        for x in 0..DIM {
+            let word_c = Self::from_quarters(&self.vec_dense_c(), None, x);
+            let rem_c = Self::from_quarters(&self.vec_remainder_c(), None, x);
+            let rot_c = Self::from_quarters(&self.vec_dense_rot_c(), None, x);
+
+            self.constrain(
+                self.is_round()
+                    * (word_c * Self::two_pow(1)
+                        - (self.quotient_c(x) * Self::two_pow(64) + rem_c.clone())),
+            );
+            self.constrain(self.is_round() * (rot_c - (self.quotient_c(x) + rem_c)));
+            self.constrain(self.is_round() * (Self::is_boolean(self.quotient_c(x))));
+
+            for q in 0..QUARTERS {
+                state_c[x][q] = self.state_a(0, x, q)
+                    + self.state_a(1, x, q)
+                    + self.state_a(2, x, q)
+                    + self.state_a(3, x, q)
+                    + self.state_a(4, x, q);
                 self.constrain(
-                    self.is_absorb()
-                        * (self.new_state(i).clone()
+                    self.is_round()
+                        * (state_c[x][q].clone()
                             - Self::from_shifts(
-                                &self.vec_sponge_shifts(),
-                                Some(i),
-                                None,
+                                &self.vec_shifts_c(),
                                 None,
                                 None,
+                                Some(x),
+                                Some(q),
                             )),
                 );
-            }
-            let sponge_shifts = self.vec_sponge_shifts();
-            for i in 0..QUARTERS * WORDS_IN_HASH {
-                // In squeeze, Check shifts correspond to the 256-bit prefix digest of the old state (current)
```

Comment:
It seems that the original code has only this `QUARTERS * WORDS_IN_HASH` squeeze loop, but the new code has additional one:
```
           // In absorb, Check shifts correspond to the decomposition of the new state
            self.constrain(
                self.is_absorb()
                    * (self.new_state(i).clone()
                        - Self::from_shifts(&self.vec_sponge_shifts(), Some(i), None, None, None)),
            );
```            
over `QUARTERS * DIM * DIM`... no?

Reply:
The squeeze loop is over `QUARTERS * WORDS_IN_HASH` but the one for absorb is over `QUARTERS * DIM * DIM`